delta-lake/src/main/java/org/apache/iceberg/delta/BaseSnapshotDeltaLakeTableAction.java

```diff
@@ -0,0 +1,370 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Type;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Takes a Delta Lake table's location and attempts to create an Iceberg table snapshot in an
+ * optional user-specified location (default to the Delta Lake table's location) with a different
+ * identifier.
+ */
+class BaseSnapshotDeltaLakeTableAction implements SnapshotDeltaLakeTable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BaseSnapshotDeltaLakeTableAction.class);
+
+  private static final String SNAPSHOT_SOURCE_PROP = "snapshot_source";
+  private static final String DELTA_SOURCE_VALUE = "delta";
+  private static final String ORIGINAL_LOCATION_PROP = "original_location";
+  private static final String PARQUET_SUFFIX = ".parquet";
+  private static final String AVRO_SUFFIX = ".avro";
+  private static final String ORC_SUFFIX = ".orc";
+  private final ImmutableMap.Builder<String, String> additionalPropertiesBuilder =
+      ImmutableMap.builder();
+  private DeltaLog deltaLog;
+  private Catalog icebergCatalog;
+  private final String deltaTableLocation;
+  private TableIdentifier newTableIdentifier;
+  private String newTableLocation;
+  private HadoopFileIO deltaLakeFileIO;
+
+  /**
+   * Snapshot a delta lake table to be an iceberg table. The action will read the delta lake table's
+   * log through the table's path, create a new iceberg table using the given icebergCatalog and
+   * newTableIdentifier, and commit all changes in one iceberg transaction.
+   *
+   * <p>The new table will only be created if the snapshot is successful.
+   *
+   * @param deltaTableLocation the delta lake table's path
+   */
+  BaseSnapshotDeltaLakeTableAction(String deltaTableLocation) {
+    this.deltaTableLocation = deltaTableLocation;
+    this.newTableLocation = deltaTableLocation;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable tableProperties(Map<String, String> properties) {
+    additionalPropertiesBuilder.putAll(properties);
+    return this;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable tableProperty(String name, String value) {
+    additionalPropertiesBuilder.put(name, value);
+    return this;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable tableLocation(String location) {
+    this.newTableLocation = location;
+    return this;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable as(TableIdentifier identifier) {
+    this.newTableIdentifier = identifier;
+    return this;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable icebergCatalog(Catalog catalog) {
+    this.icebergCatalog = catalog;
+    return this;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable deltaLakeConfiguration(Configuration conf) {
+    this.deltaLog = DeltaLog.forTable(conf, deltaTableLocation);
+    this.deltaLakeFileIO = new HadoopFileIO(conf);
+    return this;
+  }
+
+  @Override
+  public Result execute() {
+    Preconditions.checkArgument(
```

Comment:
Thank you for your suggestion. Based on my test with `delta-standalone:0.6.0`, the exception will be thrown when we try to initialize the `DeltaLog` for table with higher reader/writer version by `DeltaLog.forTable(conf, deltaTableLocation)`. 

Since we need `DeltaLog` to access table properties and version log of a delta table, It seems that additional check related to the protocol version will never be reached as the exception is thrown earlier. 

I think we can add relevant checks and the support for column mapping when newer version of `delta-standalone` is published. What do you think?

Here is a full stacktrace that copied from AWS Glue when I tried to snapshot a delta lake table using column mapping and got the `InvalidProtocolVersionException`:
```
23/01/05 00:27:50 ERROR ProcessLauncher: Exception in User Class
io.delta.standalone.internal.exception.DeltaErrors$InvalidProtocolVersionException: 
Delta protocol version (2,5) is too new for this version of Delta
Standalone Reader/Writer (1,2). Please upgrade to a newer release.

	at io.delta.standalone.internal.DeltaLogImpl.assertProtocolRead(DeltaLogImpl.scala:214) ~[delta-standalone_2.12-0.6.0.jar:0.6.0]
	at io.delta.standalone.internal.SnapshotImpl.<init>(SnapshotImpl.scala:272) ~[delta-standalone_2.12-0.6.0.jar:0.6.0]
	at io.delta.standalone.internal.SnapshotManagement.createSnapshot(SnapshotManagement.scala:257) ~[delta-standalone_2.12-0.6.0.jar:0.6.0]
	at io.delta.standalone.internal.SnapshotManagement.getSnapshotAtInit(SnapshotManagement.scala:224) ~[delta-standalone_2.12-0.6.0.jar:0.6.0]
	at io.delta.standalone.internal.SnapshotManagement.$init$(SnapshotManagement.scala:37) ~[delta-standalone_2.12-0.6.0.jar:0.6.0]
	at io.delta.standalone.internal.DeltaLogImpl.<init>(DeltaLogImpl.scala:47) ~[delta-standalone_2.12-0.6.0.jar:0.6.0]
	at io.delta.standalone.internal.DeltaLogImpl$.apply(DeltaLogImpl.scala:263) ~[delta-standalone_2.12-0.6.0.jar:0.6.0]
	at io.delta.standalone.internal.DeltaLogImpl$.forTable(DeltaLogImpl.scala:241) ~[delta-standalone_2.12-0.6.0.jar:0.6.0]
	at io.delta.standalone.internal.DeltaLogImpl.forTable(DeltaLogImpl.scala) ~[delta-standalone_2.12-0.6.0.jar:0.6.0]
	at io.delta.standalone.DeltaLog.forTable(DeltaLog.java:164) ~[delta-standalone_2.12-0.6.0.jar:0.6.0]
	at org.apache.iceberg.delta.BaseMigrateDeltaLakeTableAction.<init>(BaseMigrateDeltaLakeTableAction.java:104) ~[iceberg-spark-runtime-3.3_migrate_delta_to_iceberg.jar:?]
	at org.apache.iceberg.spark.actions.MigrateDeltaLakeTableSparkAction.<init>(MigrateDeltaLakeTableSparkAction.java:37) ~[iceberg-spark-runtime-3.3_migrate_delta_to_iceberg.jar:?]
	at org.apache.iceberg.spark.actions.SparkActions.migrateDeltaLakeTable(SparkActions.java:83) ~[iceberg-spark-runtime-3.3_migrate_delta_to_iceberg.jar:?]
	at GlueApp$.main(delta_iceberg_migration_test.scala:21) ~[delta_iceberg_migration_test.scala.jar:?]
	at GlueApp.main(delta_iceberg_migration_test.scala) ~[delta_iceberg_migration_test.scala.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_352]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_352]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_352]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_352]
	at com.amazonaws.services.glue.SparkProcessLauncherPlugin.invoke(ProcessLauncher.scala:50) ~[AWSGlueSparkResourceManager-1.0.jar:?]
	at com.amazonaws.services.glue.SparkProcessLauncherPlugin.invoke$(ProcessLauncher.scala:50) ~[AWSGlueSparkResourceManager-1.0.jar:?]
	at com.amazonaws.services.glue.ProcessLauncher$$anon$1.invoke(ProcessLauncher.scala:80) ~[AWSGlueSparkResourceManager-1.0.jar:?]
	at com.amazonaws.services.glue.ProcessLauncher.launch(ProcessLauncher.scala:145) ~[AWSGlueSparkResourceManager-1.0.jar:?]
	at com.amazonaws.services.glue.ProcessLauncher$.main(ProcessLauncher.scala:30) ~[AWSGlueSparkResourceManager-1.0.jar:?]
	at com.amazonaws.services.glue.ProcessLauncher.main(ProcessLauncher.scala) ~[AWSGlueSparkResourceManager-1.0.jar:?]
```

Reply:
>the DeltaLog for table with higher reader/writer version by DeltaLog.forTable(conf, deltaTableLocation)

if this check is embeded within the forTable itself then I think I am fine with it :)